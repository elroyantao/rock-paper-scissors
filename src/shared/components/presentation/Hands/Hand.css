.Hand {
  width: 150px;
  max-width: 19%;
  display: flex;
  justify-content: center;
  align-items: center;
  height: 80px;
}

.Hand--progress {
  composes: Hand;
  max-width: initial;
}

.Hand-icon {
  width: 50px;
  height: 50px;
  transition: width .5s, height .5s;
  border: solid 1px black;
  border-radius: 75px;
  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;
  /*max-width: 60%;*/
  /*transition: width .5s;*/
}

.Hand-icon:hover {
  width: 75px;
  height: 75px;
  /*max-width: 80%;*/
}

.Hand-icon--progress {
  composes: Hand-icon;
}
.Hand-icon--progress:hover {
  width: 50px;
  height: 50px;
}

.Hand-image {
  width: 100%;
  height: 100%;
  background-size: contain;
  background-position: center;
  background-repeat: no-repeat;
  border-radius: 50px;
}

.Hand-image--rock {
  composes: Hand-image;
  background-image: url('../../../../../assets/rock.png');
}
.Hand-image--paper {
  composes: Hand-image;
  background-image: url('../../../../../assets/paper.png');
}
.Hand-image--scissors {
  composes: Hand-image;
  background-image: url('../../../../../assets/scissors.png');
  background-size: 70%;
}
.Hand-image--lizard {
  composes: Hand-image;
  background-image: url('../../../../../assets/lizard.png');
  background-size: 75%;
  background-position-x: 75%;
}
.Hand-image--spock {
  composes: Hand-image;
  background-image: url('../../../../../assets/spock.png');
  background-size: 65%;
}

.Hand-image--classicWaiting {
  composes: Hand-image;
  background-size: 65%;
  animation: classicWait .6s infinite;
}
.Hand-image--advancedWaiting {
  composes: Hand-image;
  background-size: 65%;
  animation: advancedWait 1s infinite;
}

@keyframes advancedWait {
    0%   {
      background-image: url('../../../../../assets/rock.png');
    }
    20%   {
      background-image: url('../../../../../assets/paper.png');
    }
    40%   {
      background-image: url('../../../../../assets/scissors.png');
    }
    60%   {
      background-image: url('../../../../../assets/lizard.png');
    }
    80%   {
      background-image: url('../../../../../assets/spock.png');
    }
}

@keyframes classicWait {
    0%   {
      background-image: url('../../../../../assets/rock.png');
    }
    33%   {
      background-image: url('../../../../../assets/paper.png');
    }
    66%   {
      background-image: url('../../../../../assets/scissors.png');
    }
}
