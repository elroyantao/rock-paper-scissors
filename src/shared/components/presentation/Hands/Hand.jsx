import React, { Component } from 'react'
import PropTypes from 'prop-types'

import styles from './Hand.css'

export default class Hand extends Component {
  static propTypes = {
    type: PropTypes.oneOfType([
      PropTypes.number,
      PropTypes.string
    ]),
    isProgress: PropTypes.bool,
    onClick: PropTypes.func
  }

  static defaultProps = {
    onClick: () => {}
  }

  getButtonName(type) {
    let name
    switch (type) {
      case 0:
        name = 'rock'
        break
      case 1:
        name = 'spock'
        break
      case 2:
        name = 'paper'
        break
      case 3:
        name = 'lizard'
        break
      case 4:
        name = 'scissors'
        break
      default:
        name = type
    }
    return (
      <div className={styles[`Hand-image--${name}`]} />
    )
  }

  applyHand = () => {
    const { onClick, type } = this.props
    if (typeof type === 'string') return
    onClick(type)
  }

  render() {
    const { type, isProgress } = this.props
    return (
      <div className={styles[`Hand${isProgress ? '--progress' : ''}`]}>
        <div className={styles[`Hand-icon${isProgress ? '--progress' : ''}`]} onClick={this.applyHand}>
          {this.getButtonName(type)}
        </div>
      </div>
    )
  }
}
