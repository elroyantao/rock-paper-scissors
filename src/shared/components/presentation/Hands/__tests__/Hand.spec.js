import testComponentHelper from '../../../../lib/test-helper'
import Hand from '../Hand'


describe('<Hand /> ', () => {
  const initialState = {
    type: 0,
    isProgress: undefined,
    onClick: jest.fn()
  }
  const renderComponent = testComponentHelper(Hand)
  describe('@renders', () => {
    it('in default state when type is 0/rock', () => {
      const { wrapper, getTree } = renderComponent(initialState)
      expect(getTree())
        .toMatchSnapshot()
      expect(wrapper.find('.Hand-image--rock').length).toBe(1)
    })
    it('when type is 1/spock', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialState,
        type: 1
      })
      expect(getTree())
        .toMatchSnapshot()
      expect(wrapper.find('.Hand-image--spock').length).toBe(1)
    })
    it('when type is 1/paper', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialState,
        type: 2
      })
      expect(getTree())
        .toMatchSnapshot()
      expect(wrapper.find('.Hand-image--paper').length).toBe(1)
    })
    it('when type is 1/lizard', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialState,
        type: 3
      })
      expect(getTree())
        .toMatchSnapshot()
      expect(wrapper.find('.Hand-image--lizard').length).toBe(1)
    })
    it('when type is 1/scissors', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialState,
        type: 4
      })
      expect(getTree())
        .toMatchSnapshot()
      expect(wrapper.find('.Hand-image--scissors').length).toBe(1)
    })
    it('when type is a string', () => {
      const { wrapper, getTree } = renderComponent({
        ...initialState,
        type: 'waiting'
      })
      expect(getTree())
        .toMatchSnapshot()
      expect(wrapper.find('.Hand-image--waiting').length).toBe(1)
    })
  })
  describe('@events', () => {
    beforeEach(() => jest.clearAllMocks())
    it('onClick hand should call onClick if type is a number', () => {
      const { wrapper, instance } = renderComponent({
        ...initialState,
        type: 2
      })
      wrapper.find('.Hand-icon').simulate('click')
      expect(instance.props.onClick).toHaveBeenCalledTimes(1)
    })
    it('onClick hand should  not call onClick if type is a string', () => {
      const { wrapper, instance } = renderComponent({
        ...initialState,
        type: 'waiting'
      })
      wrapper.find('.Hand-icon').simulate('click')
      expect(instance.props.onClick).not.toHaveBeenCalled()
    })
  })
})
