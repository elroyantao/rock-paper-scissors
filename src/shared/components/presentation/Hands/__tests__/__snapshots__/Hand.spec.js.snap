// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hand />  @renders in default state when type is 0/rock 1`] = `
<div
  className="Hand"
>
  <div
    className="Hand-icon"
    onClick={[Function]}
  >
    <div
      className="Hand-image--rock"
    />
  </div>
</div>
`;

exports[`<Hand />  @renders when type is 1/lizard 1`] = `
<div
  className="Hand"
>
  <div
    className="Hand-icon"
    onClick={[Function]}
  >
    <div
      className="Hand-image--lizard"
    />
  </div>
</div>
`;

exports[`<Hand />  @renders when type is 1/paper 1`] = `
<div
  className="Hand"
>
  <div
    className="Hand-icon"
    onClick={[Function]}
  >
    <div
      className="Hand-image--paper"
    />
  </div>
</div>
`;

exports[`<Hand />  @renders when type is 1/scissors 1`] = `
<div
  className="Hand"
>
  <div
    className="Hand-icon"
    onClick={[Function]}
  >
    <div
      className="Hand-image--scissors"
    />
  </div>
</div>
`;

exports[`<Hand />  @renders when type is 1/spock 1`] = `
<div
  className="Hand"
>
  <div
    className="Hand-icon"
    onClick={[Function]}
  >
    <div
      className="Hand-image--spock"
    />
  </div>
</div>
`;

exports[`<Hand />  @renders when type is a string 1`] = `
<div
  className="Hand"
>
  <div
    className="Hand-icon"
    onClick={[Function]}
  >
    <div
      className="Hand-image--waiting"
    />
  </div>
</div>
`;
